src/Lean/Meta/Tactic/FunInd.lean

```diff
@@ -0,0 +1,855 @@
+
+```
+def ackermann : Nat → Nat → Nat
+  | 0, m => m + 1
+  | n+1, 0 => ackermann n 1
+  | n+1, m+1 => ackermann n (ackermann (n + 1) m)
+derive_induction ackermann
+```
+we get
+```
+ackermann.induct (motive : Nat → Nat → Prop) (case1 : ∀ (m : Nat), motive 0 m)
+  (case2 : ∀ (n : Nat), motive n 1 → motive (Nat.succ n) 0)
+  (case3 : ∀ (n m : Nat), motive (n + 1) m → motive n (ackermann (n + 1) m) → motive (Nat.succ n) (Nat.succ m))
+  (x x : Nat) : motive x x
+```
+
+## Spec
+
+The functional induction principle
+
+* Take the same fixed parameters as the function.
+* The motive takes the same non-fixed parameter as the original function.
+* For each branch of the original function, there is a case in the induction principle.
+  Here branch roughly corresponds to tail-call positions: Branches of top-level
+  if-then-else and matches.
+* The local context of the branch (e.g. the condition of an if-then-else) is provided as assumptions
+  in the case.
+* In addition, for every recursive call in that branch, an induction hypothesis asserting the motive
+  for the arguments of the recursive call is provided.
+* If the recursive call is under binder and it, or its proof of termination, on the the bound,
+  values, these become assumptions on the inductive hypothesis.
+* Mutual recursion is supported and results in multiple motives.
+
+
+## Implementation overview
+
+For a non-mutual, unary function `foo` (or else for the `_unary` function), we
+
+1. expect its definition, possibly after some `whnf`’ing, to be of the form
+   ```
+   def foo := fun x₁ … xₙ (y : a) => fix (fun y' oldIH => body) y
+   ```
+   where `xᵢ…` are the fixed prefix and `y` is parameter of the function.
+
+2. From this structure we derive the type of the `motive`, and start assembling the induction
+   principle:
+   ```
+   def foo.induct := fun x₁ … xₙ (motive : (y : a) → Prop) =>
+    fix (fun y' newIH => T[body])
+   ```
+
+3. The first phase, transformation `T1[body]` (implemented in) `buildInductionBody`,
+   mirrors the branching structure of `foo`, i.e. replicates `ite` and matcher applications,
+   while adjusting their motive. It also unfolds call to `oldIH` and collects induction hypotheses
+   in conditions (see below).
+
+   In particular when translating a `match`, it is prepared to see the idiom
+   as introduced by `mkFix` via `Lean.Meta.MatcherApp.addArg?`, which refines the type of `oldIH`
+   throught the match. The transformation will replace `oldIH` with `newIH` here.
+   ```
+        T[(match (motive := fun oldIH => …) y with | … => fun oldIH' => body) oldIH]
+    ==> (match (motive := fun newIH => …) y with | … => fun newIH' => T[body]) newIH
+   ```
+
+4. When a tail position (no more branching) is found, function `buildInductionCase` assembles the
+   type of the case: A fresh `MVar` asserts the current goal, unwanted values from the local context
+   are cleared, and the current `body` is frisked for induction hypotheses using `collectIHs`,
```

Comment:
What does "frisked" mean here? I only know the term as searching someone's clothes for hidden weapons, which seems unlikely to be what was meant!

Reply:
I didn't think frisking is reservd for weapons.
searched through or combed are maybe less flowery phrasings :-)

Reply:
```suggestion
   are cleared, and the current `body` is combed for induction hypotheses using `collectIHs`,
```

Reply:
Oh, "frisked" could work here, the sudden metaphor just threw me off and made me think it was a mistake rather than intentional. For me, "comb" has more a connotation of "setting in place" than "removing the dirt", although the latter certainly does work in the case of lice or sand. 

What about "the current `body`'s induction hypotheses are confiscated by `collectIHs`" if you want the metaphor?

But I think the metaphor was jarring in a body of text that's otherwise straightforward and technical, so I'd reduce the fun here if anything.